internal/api/v1/application/stage.go

```diff
@@ -339,50 +340,153 @@ func validateBlob(ctx context.Context, blobUID string, app models.AppRef, s3Conn
 	return nil
 }
 
-// newPipelineRun is a helper which creates a Tekton pipeline run
-// resource from the given staging params
-func newPipelineRun(app stageParam) *v1beta1.PipelineRun {
-	str := v1beta1.NewArrayOrString
+// newJobRun is a helper which creates the Job related resources from
+// the given staging params. That is the job itself, and a secret
+// holding the job's environment. Which is a copy of the app
+// environment + standard variables.
+func newJobRun(app stageParam) (*batchv1.Job, *corev1.Secret) {
+
+	// fake stage params of the previous to pull the old image url from.
+	previous := app
+	previous.Stage = models.NewStage(app.PreviousStageID)
 
 	protocol := "http"
 	if app.S3ConnectionDetails.UseSSL {
 		protocol = "https"
 	}
-	awsScript := fmt.Sprintf("aws --endpoint-url %s://$1 s3 cp s3://$2/$3 $(workspaces.source.path)/$3", protocol)
-	awsArgs := []string{
-		app.S3ConnectionDetails.Endpoint,
-		app.S3ConnectionDetails.Bucket,
-		app.BlobUID,
+
+	awsScript := fmt.Sprintf("echo Extracting _ _ __ ___ _____ $(whoami) ; aws --endpoint-url %s://%s s3 cp s3://%s/%s /workspace/source/%s ; echo _ _ __ ___ _____ Extracted",
+		protocol, app.S3ConnectionDetails.Endpoint, app.S3ConnectionDetails.Bucket,
+		app.BlobUID, app.BlobUID)
+
+	unpackScript := fmt.Sprintf("echo Unpacking _ _ __ ___ _____ $(whoami) ; mkdir /workspace/source/app ; tar -xvf /workspace/source/%s -C /workspace/source/app/; rm /workspace/source/%s; chown -R 1000:1000 /workspace ; cp -vL /platform/appenv/* /platform/env/ ; ls -la /platform/env ; echo _ _ __ ___ _____ Unpacked", app.BlobUID, app.BlobUID)
+
+	buildpackScript := fmt.Sprintf("echo Creating _ _ __ ___ _____ $(whoami) ; ls -la /platform/env ; /cnb/lifecycle/creator -app=/workspace/source/app -cache-dir=/workspace/cache -uid=1000 -gid=1000 -layers=/layers -platform=/platform -report=/layers/report.toml -process-type=web -skip-restore=false -previous-image=%s %s && ( curl -X POST http://localhost:4191/shutdown || true )",
```

Comment:
I believe it's better to have these scripts defined in a ConfigMap and mounted on the Job. It will make them more readable. Also by not embedding them in the binary, it will make it possible to change them without building a new binary (e.g. to debug or change the behavior). 

By the way, I think `|| true` deserves a comment. Something like `curl localhost:4191 is used to stop the linkerd proxy container gracefully. We use "|| true" in case linkerd is not deployed`.

Reply:
Feel free to create a separate issue for the extraction of these scripts to a ConfigMap if you agree in doing this.

Reply:
Will do the comment. Move to ConfigMaps will be TODO's, and get a separate issue.

Reply:
New issue #1175

Reply:
Done. Commit 7eb20b7e67e5b4cff825d27bb2309b3486e1cc68.